### PR TITLE
Additional guild features & initial guild DB implementation

### PR DESCRIPTION
**Pull request by Zintixx**
This includes more features of Guilds:

- Applying to guild
- Upgrading buffs
- Upgrading guild services
- Changing and upgrading guild housing

What's currently missing:

- Guild raids (of course)
- Guild mini games
- Guild gift bank/log
- Correctly spawning guild services to your own proper guild house (maybe needs to wait for instance implementation?)

That said, proper parsing for these items have been added.

Do note that at the moment, there is database implementation for guilds but it's very minimal. Currently it handles inviting, creating, and disbanding guilds. More to follow as sessions get reworked. It's kinda wonky debugging this in the midst of it.

## Misc
I've changed the `GameCommandAction` so you'll need to begin your command with a forward slash (/). This is nicer than having to shout every command into the field

#83

### CODE REVIEW

File: MapleServer2/PacketHandlers/Game/GuildHandler.cs [hunk 1]
```diff
@@ -217,33 +319,27 @@ private static void HandleInviteResponse(GameSession session, PacketReader packe
...
 
             inviter.Session.Send(GuildPacket.InviteNotification(inviteeName, response));
             session.Send(GuildPacket.InviteResponseConfirm(inviter, session.Player, guild, response));
-            session.FieldManager.BroadcastPacket(GuildPacket.UpdateGuildTag(session.Player, guildName));
+            session.FieldManager.BroadcastPacket(GuildPacket.UpdateGuildTag2(session.Player, guildName));
             guild.AddMember(session.Player);
-            session.Player.GuildId = guild.Id;
-            guild.BroadcastPacketGuild(GuildPacket.MemberBroadcastJoinNotice(session.Player, inviterName, response, rank));
-            guild.BroadcastPacketGuild(GuildPacket.MemberJoin(session.Player));
+            GuildMember member = guild.Members.FirstOrDefault(x => x.Player == session.Player);
```
Comment (AngeloTadeucci) [inline MapleServer2/PacketHandlers/Game/GuildHandler.cs:327]: It's needed to check if member is null?

File: MapleServer2/Servers/Game/GameServer.cs [hunk 2]
```diff
@@ -22,6 +27,17 @@ public class GameServer : Server<GameSession>
 
         public void Start()
         {
+            List<Guild> guilds = new List<Guild>();
+            using (DatabaseContext context = new DatabaseContext())
+            {
+                guilds = context.Guilds
+                .Include(p => p.Members).ThenInclude(p => p.Player).ThenInclude(p => p.Levels)
+                .Include(p => p.Leader).ToList();
+            }
```
Comment (AngeloTadeucci) [inline MapleServer2/Servers/Game/GameServer.cs:36]: Also move this to `DatabaseManager` please.

File: MapleServer2/Types/Guild/Guild.cs [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+                if (Funds >= property.FundMax)
+                {
+                    return;
+                }
+                Funds += funds;
+
+                if (Funds >= property.FundMax)
+                {
+                    Funds = property.FundMax;
+                }
```
Comment (AngeloTadeucci) [inline MapleServer2/Types/Guild/Guild.cs:131]: Math.Min

File: MapleServer2/Types/Player.cs [hunk 4]
```diff
@@ -109,9 +109,9 @@ public class Player
         public int[] GroupChatId;
 
         // TODO: Rework to use Class Guild
-        public long GuildId;
-        public string GuildName;
-        public int GuildContribution;
+        public Guild Guild;
+        public GuildMember GuildMember;
+        public List<GuildApplication> GuildApplications = new List<GuildApplication>();
```
Comment (AngeloTadeucci) [inline MapleServer2/Types/Player.cs:114]: Have you tested if this initialization will overwrite the return from the database?

File: MapleServer2/Database/DatabaseManager.cs [hunk 5]
```diff
@@ -132,6 +132,8 @@ public static void UpdateCharacter(Player player)
                 context.Entry(player.Levels).State = EntityState.Modified;
                 context.Entry(player.BankInventory).State = EntityState.Modified;
                 context.Entry(player.Inventory).State = EntityState.Modified;
+                context.Entry(player.Guild).State = EntityState.Modified;
+                context.Entry(player.GuildMember).State = EntityState.Modified;
```
Comment (AngeloTadeucci) [inline MapleServer2/Database/DatabaseManager.cs:136]: Check if guild and guildmember is null

**Review by AngeloTadeucci — CHANGES_REQUESTED** [review]

File: MapleServer2/Packets/CharacterListPacket.cs [hunk 6]
```diff
@@ -3,6 +3,7 @@
 using MaplePacketLib2.Tools;
 using MapleServer2.Constants;
 using MapleServer2.Packets.Helpers;
+using MapleServer2.Servers.Game;
```
Comment (AngeloTadeucci) [inline MapleServer2/Packets/CharacterListPacket.cs:6]: Not being used

**Review by AngeloTadeucci — CHANGES_REQUESTED** [review]

File: MapleServer2/Types/Guild/Guild.cs (hunk 3 above)
Comment (Zintixx) [reply MapleServer2/Types/Guild/Guild.cs:131]: :thonk:

File: MapleServer2/Types/Player.cs (hunk 4 above)
Comment (Zintixx) [reply MapleServer2/Types/Player.cs:114]: err?? I'm not sure. I've restarted the server a few times and havent had any overwrite happen. if thats what you mean

File: MapleServer2/Types/Guild/Guild.cs (hunk 3 above)
Comment (AngeloTadeucci) [reply MapleServer2/Types/Guild/Guild.cs:131]: Math.Min(x, y) gets the values which has the lowest value, exactly what you want here

File: MapleServer2/Types/Player.cs (hunk 4 above)
Comment (AngeloTadeucci) [reply MapleServer2/Types/Player.cs:114]: I checked and it's ok!

**Review by AngeloTadeucci — APPROVED** [review]

**Review by AlanMorel — CHANGES_REQUESTED** [review]

**Review by AngeloTadeucci — APPROVED** [review]

**Pull request merged**